jobs/gtfs-rt-parser/gtfs_rt_parser.py

```diff
@@ -123,109 +218,440 @@ def identify_files(glob, rt_file_type: RTFileType, progress=False) -> List[RTFil
+            gzip_fname=str(tmp_dir + f"/errors_{hour.suffix}" + JSONL_GZIP_EXTENSION),
+            out_path=f"{hour.errors_hive_path}{JSONL_GZIP_EXTENSION}",
+            records=errors,
+            pbar=pbar,
+            verbose=verbose,
+        )
+    else:
+        log(
+            f"WARNING: no records at all for {hour.errors_hive_path}",
+            fg=typer.colors.YELLOW,
+            pbar=pbar,
+        )
 
-def try_handle_one_feed(*args, **kwargs) -> Optional[Exception]:
-    try:
-        parse_file(*args, **kwargs)
-    except Exception as e:
-        typer.echo(
-            f"got exception while handling feed: {str(e)} from {str(e.__cause__ or e.__context__)} {traceback.format_exc()}",
-            err=True,
+
+# Originally this whole function was retried, but tmpdir flakiness will throw
+# exceptions in backoff's context, which ruins things
+def parse_and_aggregate_hour(
+    hour: RTAggregation,
+    jar_path: Path,
+    tmp_dir: str,
+    verbose: bool = False,
+    parse: bool = False,
+    validate: bool = False,
+    pbar=None,
+):
+    if not parse and not validate:
+        raise ValueError("skipping both parsing and validation does nothing for us!")
+
+    fs = get_fs()
+    suffix = hour.suffix
+    dst_path_gtfs = f"{tmp_dir}/gtfs_{suffix}/"
+    os.mkdir(dst_path_gtfs)
+    dst_path_rt = f"{tmp_dir}/rt_{suffix}/"
+    get_with_retry(
+        fs,
+        rpath=[file.path for file in hour.source_files],
+        lpath=[
+            os.path.join(dst_path_rt, file.timestamped_filename)
+            for file in hour.source_files
+        ],
+    )
+
+    if validate:
+        validate_and_upload(
+            fs=fs,
+            jar_path=jar_path,
+            dst_path_gtfs=dst_path_gtfs,
+            dst_path_rt=dst_path_rt,
+            tmp_dir=tmp_dir,
+            hour=hour,
+            verbose=verbose,
+            pbar=pbar,
+        )
+
+    if parse:
+        parse_and_upload(
+            fs=fs,
+            dst_path_rt=dst_path_rt,
+            tmp_dir=tmp_dir,
+            hour=hour,
+            verbose=verbose,
+            pbar=pbar,
         )
-        # a lot of these are thrown by tmpdir, and are potentially flakes; but if they were thrown during
-        # handling of another exception, we want that exception still
-        if isinstance(e, OSError) and e.errno == 39:
-            return e.__cause__ or e.__context__
-        return e
 
 
 def main(
     file_type: RTFileType,
     glob: str,
-    dst_bucket=get_bucket(),
+    dst_bucket: str,
     limit: int = 0,
+    parse: bool = False,
+    validate: bool = False,
     progress: bool = typer.Option(
         False,
         help="If true, display progress bar; useful for development but not in production.",
     ),
+    verbose: bool = False,
     threads: int = 4,
+    jar_path: Path = JAR_DEFAULT,
 ):
-    typer.secho(f"Parsing {file_type} from {glob}", fg=typer.colors.MAGENTA)
+    assert parse ^ validate, "must either parse xor validate"
```

Comment:
**question**: you've got that other check above for if not parse or validate -- do we need to check in both places?

Reply:
This is a separate check; the check in `main()` is to assert the job is called with either parse or validate, as a safety check for Airflow.

The underlying `parse_and_aggregate_hour()` could technically do both in the future if we want.